src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1096 @@
+
+		UpdateBackgroundTask(task);
+		UpdateDependingTasks(task);
+		UpdateBackgroundJob(task->jobid);
+
+		dsm_detach(seg);
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+	}
+
+	MemoryContextSwitchTo(oldContextPerJob);
+	MemoryContextDelete(perTaskContext);
+}
+
+
+#if PG_VERSION_NUM < PG_VERSION_15
+static const char *
+error_severity(int elevel)
+{
+	const char *prefix;
+
+	switch (elevel)
+	{
+		case DEBUG1:
+		case DEBUG2:
+		case DEBUG3:
+		case DEBUG4:
+		case DEBUG5:
+		{
+			prefix = gettext_noop("DEBUG");
+			break;
+		}
+
+		case LOG:
+		case LOG_SERVER_ONLY:
+		{
+			prefix = gettext_noop("LOG");
+			break;
+		}
+
+		case INFO:
+		{
+			prefix = gettext_noop("INFO");
+			break;
+		}
+
+		case NOTICE:
+		{
+			prefix = gettext_noop("NOTICE");
+			break;
+		}
+
+		case WARNING:
+		{
+			prefix = gettext_noop("WARNING");
+			break;
+		}
+
+#if PG_VERSION_NUM >= PG_VERSION_14
+		case WARNING_CLIENT_ONLY:
+		{
+			prefix = gettext_noop("WARNING");
+			break;
+		}
+#endif
+
+		case ERROR:
+		{
+			prefix = gettext_noop("ERROR");
+			break;
+		}
+
+		case FATAL:
+		{
+			prefix = gettext_noop("FATAL");
+			break;
+		}
+
+		case PANIC:
+		{
+			prefix = gettext_noop("PANIC");
+			break;
+		}
+
+		default:
+		{
+			prefix = "???";
+			break;
+		}
+	}
+
+	return prefix;
+}
+
+
+#endif
+
+
+/*
+ * bgw_generate_returned_message -
+ *      generates the message to be inserted into the job_run_details table
+ *      first part is comming from error_severity (elog.c)
+ */
+static void
+bgw_generate_returned_message(StringInfoData *display_msg, ErrorData edata)
+{
+	const char *prefix = error_severity(edata.elevel);
+	appendStringInfo(display_msg, "%s: %s", prefix, edata.message);
+	if (edata.detail != NULL)
+	{
+		appendStringInfo(display_msg, "\nDETAIL: %s", edata.detail);
+	}
+
+	if (edata.hint != NULL)
+	{
+		appendStringInfo(display_msg, "\nHINT: %s", edata.hint);
+	}
+
+	if (edata.context != NULL)
+	{
+		appendStringInfo(display_msg, "\nCONTEXT: %s", edata.context);
+	}
+}
+
+
+static void
+UpdateDependingTasks(BackgroundTask *task)
+{
+	switch (task->status)
+	{
+		case BACKGROUND_TASK_STATUS_DONE:
+		{
+			UnblockDependingBackgroundTasks(task->jobid, task->taskid);
+			break;
+		}
+
+		case BACKGROUND_TASK_STATUS_ERROR:
+		{
+			/* when we error this task, we need to unschedule all dependant tasks */
+			UnscheduleDependantTasks(task->jobid, task->taskid);
+			break;
+		}
+
+		default:
+		{
+			/* nothing to do for other states */
+			break;
+		}
+	}
+}
+
+
+static void
+ConsumeTaskWorkerOutput(shm_mq_handle *responseq, BackgroundTask *task, bool *hadError,
+						bool *cancelled)
+{
+	/*
+	 * Message-parsing routines operate on a null-terminated StringInfo,
+	 * so we must construct one.
+	 */
+	StringInfoData msg = { 0 };
+	initStringInfo(&msg);
+
+	for (;;)
+	{
+		resetStringInfo(&msg);
+
+		/* Get next message. */
+		Size nbytes = 0;
+		void *data = NULL;
+		shm_mq_result res = shm_mq_receive(responseq, &nbytes, &data, false);
```

Comment:
would be good to create a variable for the last argument, since it shows that we block here